packages/aws-cdk/lib/api/garbage-collection/stack-refresh.ts

```diff
@@ -0,0 +1,180 @@
+import { CloudFormation } from 'aws-sdk';
+import { sleep } from '../../../test/util';
+import { debug } from '../../logging';
+
+export class ActiveAssetCache {
+  private readonly stacks: Set<string> = new Set();
+
+  public rememberStack(stackTemplate: string) {
+    this.stacks.add(stackTemplate);
+  }
+
+  public contains(asset: string): boolean {
+    for (const stack of this.stacks) {
+      if (stack.includes(asset)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | undefined>) {
+  let finished = false;
+  let nextToken: string | undefined;
+  while (!finished) {
+    nextToken = await cb(nextToken);
+    if (nextToken === undefined) {
+      finished = true;
+    }
+  }
+}
+
+/** We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+ * If we encounter this status, we will wait up to the maxWaitTime before erroring out
+ */
+async function listStacksNotBeingReviewed(cfn: CloudFormation, maxWaitTime: number, nextToken: string | undefined) {
+  let sleepMs = 500;
+
+  while (sleepMs < maxWaitTime) {
```

Comment:
This is a weird condition.

"I will continue until my next sleep would have been 10 minutes"

I think one of the following equivalent conditions would have been more readable:

* Exactly N retries (leads to the same end condition, but more understandable)
* Wait until at least X seconds have elapsed 

The name is also weird. `maxWaitTime` sort of implies you're willing to wait "at most X seconds", but that's not what this condition does. It waits at most `X * 2` seconds, because it will also have waited all the intervals leading up to the interval that is this large.

The simplest and most readable solution for a retry is to use a deadline. (Bonus points that it's trivial to replace a deadline with a cancellation token should you ever want/need to).


```ts
const deadline = Date.now() + SOME_TIME;
while (Date.now() <= deadline) {
  // Do stuff
}
```

Reply:
thats a much more elegant solution